httpclient/client.go

```diff
@@ -120,25 +120,29 @@ func (c *Client) Delete(url string, headers http.Header) (*http.Response, error)
 func (c *Client) Do(request *http.Request) (*http.Response, error) {
 	request.Close = true
 
-	var reqBuffer []byte
-
-	if request != nil && request.Body != nil {
-		var err error
-
-		// Storing request buffer to create new reader on each request
-		reqBuffer, err = ioutil.ReadAll(request.Body)
+	var bodyReader *bytes.Reader
 
+	if request.Body != nil {
+		reqData, err := ioutil.ReadAll(request.Body)
 		if err != nil {
 			return nil, err
 		}
+		bodyReader = bytes.NewReader(reqData)
+		request.Body = ioutil.NopCloser(bodyReader) // prevents closing the body between retries
 	}
+
 	multiErr := &valkyrie.MultiError{}
 	var response *http.Response
 
 	for i := 0; i <= c.retryCount; i++ {
 		var err error
-		request.Body = ioutil.NopCloser(bytes.NewBuffer(reqBuffer))
```

Comment:
@dcu Do we need Seek, wouldn't just setting `request.Body` work in the next retry ? The solution which was there before. Is this to prevent multiple allocations ?

Reply:
hi, yes it prevents multiple allocations and the behavior is the same. I added tests and benchmarks to test that

Reply:
do you have any concern?

Reply:
@dcu No concerns as such, was curious on how does this add benefit. Cool not allocating everytime makes sense. Thanks